File: matlab-package/BoardShim.m
classdef BoardShim
    properties
        libname
        board_id
        port_name
    end
    methods
        function obj = BoardShim (board_id, port_name)
            if ispc
                obj.libname = 'BoardController';
                if not (libisloaded ('BoardController'))
                    loadlibrary ('./lib/BoardController.dll', './inc/board_controller.h', 'addheader', './inc/board_info_getter.h');
                end
            elseif ismac
                obj.libname = 'libBoardController';
                if not (libisloaded ('libBoardController'))
                    loadlibrary ('./lib/libBoardController.dylib', './inc/board_controller.h', 'addheader', './inc/board_info_getter.h');
                end
            elseif isunix
                obj.libname = 'libBoardController';
                if not (libisloaded ('libBoardController'))
                    loadlibrary ('./lib/libBoardController.so', './inc/board_controller.h', 'addheader', './inc/board_info_getter.h');
                end
            else
                error ('OS not supported!')
            end
            obj.port_name = port_name;
            obj.board_id = int32 (board_id);
        end

        function check_ec (obj, ec, task_name)
            if (ExitCodes.STATUS_OK.ne(ec))
                error ('%s non zero ec: %d', task_name, ec)
            end
        end

        function prepare_session (obj)
            task_name = 'prepare_session';
            exit_code = calllib (obj.libname, task_name, obj.board_id, obj.port_name);
            obj.check_ec (exit_code, task_name);
        end

        function config_board (obj, config)
            task_name = 'config_board';
            exit_code = calllib (obj.libname, task_name, config, obj.board_id, obj.port_name);
            obj.check_ec (exit_code, task_name);
        end

        function start_stream (obj, buffer_size)
            task_name = 'start_stream';
            exit_code = calllib (obj.libname, task_name, buffer_size, obj.board_id, obj.port_name);
            obj.check_ec (exit_code, task_name);
        end

        function stop_stream (obj)
            task_name = 'stop_stream';
            exit_code = calllib (obj.libname, task_name, obj.board_id, obj.port_name);
            obj.check_ec (exit_code, task_name);
        end

        function release_session (obj)
            task_name = 'release_session';
            exit_code = calllib (obj.libname, task_name, obj.board_id, obj.port_name);
            obj.check_ec (exit_code, task_name);
        end

        function set_log_level (obj, log_level)
            task_name = 'set_log_level';
            exit_code = calllib (obj.libname, task_name, log_level);
            obj.check_ec (exit_code, task_name);
        end

        function set_log_file (obj, log_file)
            task_name = 'set_log_file';
            exit_code = calllib (obj.libname, task_name, log_file);
            obj.check_ec (exit_code, task_name);
        end

        function enable_board_logger (obj)
            obj.set_log_level (2)
        end

        function enable_dev_board_logger (obj)
            obj.set_log_level (0)
        end

        function disable_board_logger (obj)
            obj.set_log_level (6)
        end

        function sampling_rate = get_sampling_rate (obj)
            task_name = 'get_sampling_rate';
            res = libpointer ('int32Ptr', 0);
            exit_code = calllib (obj.libname, task_name, obj.board_id, res);
            obj.check_ec (exit_code, task_name);
            sampling_rate = res.Value;
        end

        function package_num_channel = get_package_num_channel (obj)
            task_name = 'get_package_num_channel';
            res = libpointer ('int32Ptr', 0);
            exit_code = calllib (obj.libname, task_name, obj.board_id, res);
            obj.check_ec (exit_code, task_name);
            package_num_channel = res.Value;
        end

        function timestamp_channel = get_timestamp_channel (obj)
            task_name = 'get_timestamp_channel';
            res = libpointer ('int32Ptr', 0);
            exit_code = calllib (obj.libname, task_name, obj.board_id, res);
            obj.check_ec (exit_code, task_name);
            timestamp_channel = res.Value;
        end

        function num_rows = get_num_rows (obj)
            task_name = 'get_num_rows';
            res = libpointer ('int32Ptr', 0);
            exit_code = calllib (obj.libname, task_name, obj.board_id, res);
            obj.check_ec (exit_code, task_name);
            num_rows = res.Value;
        end

        function num_data_point = get_board_data_count (obj)
            task_name = 'get_board_data_count';
            data_count = libpointer ('int32Ptr', 0);
            exit_code = calllib (obj.libname, task_name, data_count, obj.board_id, obj.port_name);
            obj.check_ec (exit_code, task_name);
            num_data_point = data_count.value;
        end

        function eeg_channels = get_eeg_channels (obj)
            task_name = 'get_emg_channels';
            len = libpointer ('int32Ptr', 0);
            channels = libpointer ('int32Ptr', zeros (1, 256));
            exit_code = calllib (obj.libname, task_name, obj.board_id, channels, len);
            obj.check_ec (exit_code, task_name);
            eeg_channels = channels.Value (1, 1:len.Value);
        end

        function emg_channels = get_emg_channels (obj)
            task_name = 'get_emg_channels';
            len = libpointer ('int32Ptr', 0);
            channels = libpointer ('int32Ptr', zeros (1, 256));
            exit_code = calllib (obj.libname, task_name, obj.board_id, channels, len);
            obj.check_ec (exit_code, task_name);
            emg_channels = channels.Value (1, 1:len.Value);
        end

        function ecg_channels = get_ecg_channels (obj)
            task_name = 'get_ecg_channels';
            len = libpointer ('int32Ptr', 0);
            channels = libpointer ('int32Ptr', zeros (1, 256));
            exit_code = calllib (obj.libname, task_name, obj.board_id, channels, len);
            obj.check_ec (exit_code, task_name);
            ecg_channels = channels.Value (1, 1:len.Value);
        end

        function eog_channels = get_eog_channels (obj)
            task_name = 'get_eog_channels';
            len = libpointer ('int32Ptr', 0);
            channels = libpointer ('int32Ptr', zeros (1, 256));
            exit_code = calllib (obj.libname, task_name, obj.board_id, channels, len);
            obj.check_ec (exit_code, task_name);
            eog_channels = channels.Value (1, 1:len.Value);
        end

        function eda_channels = get_eda_channels (obj)
            task_name = 'get_eda_channels';
            len = libpointer ('int32Ptr', 0);
            channels = libpointer ('int32Ptr', zeros (1, 256));
            exit_code = calllib (obj.libname, task_name, obj.board_id, channels, len);
            obj.check_ec (exit_code, task_name);
            eda_channels = channels.Value (1, 1:len.Value);
        end

        function ppg_channels = get_ppg_channels (obj)
            task_name = 'get_eda_channels';
            len = libpointer ('int32Ptr', 0);
            channels = libpointer ('int32Ptr', zeros (1, 256));
            exit_code = calllib (obj.libname, task_name, obj.board_id, channels, len);
            obj.check_ec (exit_code, task_name);
            ppg_channels = channels.Value (1, 1:len.Value);
        end

        function accel_channels = get_accel_channels (obj)
            task_name = 'get_eda_channels';
            len = libpointer ('int32Ptr', 0);
            channels = libpointer ('int32Ptr', zeros (1, 256));
            exit_code = calllib (obj.libname, task_name, obj.board_id, channels, len);
            obj.check_ec (exit_code, task_name);
            accel_channels = channels.Value (1, 1:len.Value);
        end

        function gyro_channels = get_gyro_channels (obj)
            task_name = 'get_eda_channels';
            len = libpointer ('int32Ptr', 0);
            channels = libpointer ('int32Ptr', zeros (1, 256));
            exit_code = calllib (obj.libname, task_name, obj.board_id, channels, len);
            obj.check_ec (exit_code, task_name);
            gyro_channels = channels.Value (1, 1:len.Value);
        end

        function other_channels = get_other_channels (obj)
            task_name = 'get_eda_channels';
            len = libpointer ('int32Ptr', 0);
            channels = libpointer ('int32Ptr', zeros (1, 256));
            exit_code = calllib (obj.libname, task_name, obj.board_id, channels, len);
            obj.check_ec (exit_code, task_name);
            other_channels = channels.Value (1, 1:len.Value);
        end

        function data_buf = get_board_data (obj)
            task_name = 'get_board_data';
            data_count = obj.get_board_data_count ();
            num_rows = obj.get_num_rows();
            data = libpointer ('doublePtr', zeros (1, data_count * num_rows));
            exit_code = calllib (obj.libname, task_name, data_count, data, obj.board_id, obj.port_name);
            obj.check_ec (exit_code, task_name);
            data_buf = transpose(reshape (data.Value, [data_count, num_rows]));
        end

        function data_buf = get_current_board_data (obj, num_samples)
            task_name = 'get_current_board_data';
            data_count = libpointer ('int32Ptr', 0);
            num_rows = obj.get_num_rows();
            data = libpointer ('doublePtr', zeros (1, num_samples * num_rows));
            exit_code = calllib (obj.libname, task_name, num_samples, data, data_count, obj.board_id, obj.port_name);
            obj.check_ec (exit_code, task_name);
            data_buf = transpose (reshape (data.Value (1,1:data_count.Value * num_rows), [data_count.Value, num_rows]));
        end
    end
end

File: csharp-package/brainflow/brainflow/board_shim.cs
using System;

using Accord.Math;


namespace brainflow
{

    public class BoardShim
    {
        public int board_id;
        public string port_name;

        public BoardShim (int board_id, string port_name)
        {
            this.board_id = board_id;
            this.port_name = port_name;
        }

        public static int get_sampling_rate (int board_id)
        {
            int[] val = new int[1];
            int res = BoardControllerLibrary.get_sampling_rate (board_id, val);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            return val[0];
        }

        public static int get_package_num_channel (int board_id)
        {
            int[] val = new int[1];
            int res = BoardControllerLibrary.get_package_num_channel (board_id, val);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            return val[0];
        }

        public static int get_timestamp_channel (int board_id)
        {
            int[] val = new int[1];
            int res = BoardControllerLibrary.get_timestamp_channel (board_id, val);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            return val[0];
        }

        public static int get_num_rows (int board_id)
        {
            int[] val = new int[1];
            int res = BoardControllerLibrary.get_num_rows (board_id, val);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            return val[0];
        }

        public static int[] get_eeg_channels (int board_id)
        {
            int[] len = new int[1];
            int[] channels = new int[512];
            int res = BoardControllerLibrary.get_eeg_channels (board_id, channels, len);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            int[] result = new int[len[0]];
            for (int i = 0; i < len[0]; i++)
            {
                result[i] = channels[i];
            }
            return result;
        }

        public static int[] get_emg_channels (int board_id)
        {
            int[] len = new int[1];
            int[] channels = new int[512];
            int res = BoardControllerLibrary.get_emg_channels (board_id, channels, len);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            int[] result = new int[len[0]];
            for (int i = 0; i < len[0]; i++)
            {
                result[i] = channels[i];
            }
            return result;
        }

        public static int[] get_ecg_channels (int board_id)
        {
            int[] len = new int[1];
            int[] channels = new int[512];
            int res = BoardControllerLibrary.get_ecg_channels (board_id, channels, len);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            int[] result = new int[len[0]];
            for (int i = 0; i < len[0]; i++)
            {
                result[i] = channels[i];
            }
            return result;
        }

        public static int[] get_eog_channels (int board_id)
        {
            int[] len = new int[1];
            int[] channels = new int[512];
            int res = BoardControllerLibrary.get_eog_channels (board_id, channels, len);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            int[] result = new int[len[0]];
            for (int i = 0; i < len[0]; i++)
            {
                result[i] = channels[i];
            }
            return result;
        }

        public static int[] get_eda_channels (int board_id)
        {
            int[] len = new int[1];
            int[] channels = new int[512];
            int res = BoardControllerLibrary.get_eda_channels (board_id, channels, len);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            int[] result = new int[len[0]];
            for (int i = 0; i < len[0]; i++)
            {
                result[i] = channels[i];
            }
            return result;
        }

        public static int[] get_ppg_channels (int board_id)
        {
            int[] len = new int[1];
            int[] channels = new int[512];
            int res = BoardControllerLibrary.get_ppg_channels (board_id, channels, len);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            int[] result = new int[len[0]];
            for (int i = 0; i < len[0]; i++)
            {
                result[i] = channels[i];
            }
            return result;
        }

        public static int[] get_accel_channels (int board_id)
        {
            int[] len = new int[1];
            int[] channels = new int[512];
            int res = BoardControllerLibrary.get_accel_channels (board_id, channels, len);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            int[] result = new int[len[0]];
            for (int i = 0; i < len[0]; i++)
            {
                result[i] = channels[i];
            }
            return result;
        }

        public static int[] get_gyro_channels (int board_id)
        {
            int[] len = new int[1];
            int[] channels = new int[512];
            int res = BoardControllerLibrary.get_gyro_channels (board_id, channels, len);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            int[] result = new int[len[0]];
            for (int i = 0; i < len[0]; i++)
            {
                result[i] = channels[i];
            }
            return result;
        }

        public static int[] get_other_channels (int board_id)
        {
            int[] len = new int[1];
            int[] channels = new int[512];
            int res = BoardControllerLibrary.get_other_channels (board_id, channels, len);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
            int[] result = new int[len[0]];
            for (int i = 0; i < len[0]; i++)
            {
                result[i] = channels[i];
            }
            return result;
        }

        public static void enable_board_logger ()
        {
            int res = BoardControllerLibrary.set_log_level (2);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
        }

        public static void disable_board_logger ()
        {
            int res = BoardControllerLibrary.set_log_level (6);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
        }

        public static void enable_dev_board_logger ()
        {
            int res = BoardControllerLibrary.set_log_level (0);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
        }

        public static void set_log_file (string log_file)
        {
            int res = BoardControllerLibrary.set_log_file (log_file);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
        }

        public void prepare_session ()
        {
            int res = BoardControllerLibrary.prepare_session (board_id, port_name);
            if (res != (int) CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
        }

        public void config_board (string config)
        {
            int res = BoardControllerLibrary.config_board (config, board_id, port_name);
            if (res != (int)CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
        }

        public void start_stream (int num_samples = 3600 * 250)
        {
            int res = BoardControllerLibrary.start_stream (num_samples, board_id, port_name);
            if (res != (int) CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
        }

        public void stop_stream ()
        {
            int res = BoardControllerLibrary.stop_stream (board_id, port_name);
            if (res != (int) CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
        }

        public void release_session ()
        {
            int res = BoardControllerLibrary.release_session (board_id, port_name);
            if (res != (int) CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (res);
            }
        }

        public int get_board_data_count ()
        {
            int[] res = new int[1];
            int ec = BoardControllerLibrary.get_board_data_count (res, board_id, port_name);
            if (ec != (int) CustomExitCodes.STATUS_OK)
            {
                throw new BrainFlowExceptioin (ec);
            }
            return res[0];
        }

        public double[,] get_current_board_data (int num_samples)
        {
            int num_rows = BoardShim.get_num_rows (board_id); 
            double[] data_arr = new double[num_samples * num_rows];
            int[] current_size = new int[1];
            int ec = BoardControllerLibrary.get_current_board_data (num_samples, data_arr, current_size, board_id, port_name);
		    if (ec != (int) CustomExitCodes.STATUS_OK) {
			    throw new BrainFlowExceptioin (ec);
            }
            double[,] result = new double[num_rows, current_size[0]];
            for (int i = 0; i < num_rows; i++)
            {
                for (int j = 0; j < current_size[0]; j++)
                {
                    result[i, j] = data_arr[i * current_size[0] + j];
                }
            }
            return result;
	    }

        public double[,] get_immediate_board_data ()
        {
		        return get_current_board_data (0);
        }

        public double[,] get_board_data ()
        {
		    int size = get_board_data_count ();
            int num_rows = BoardShim.get_num_rows (board_id);
            double[] data_arr = new double[size * num_rows];
            int ec = BoardControllerLibrary.get_board_data (size, data_arr, board_id, port_name);
		    if (ec != (int) CustomExitCodes.STATUS_OK) {
                throw new BrainFlowExceptioin (ec);
            }
            double[,] result = new double[num_rows, size];
            for (int i = 0; i < num_rows; i++)
            {
                for (int j = 0; j < size; j++)
                {
                    result[i, j] = data_arr[i * size + j];
                }
            }
            return result;
        }
    }
}
